rllib/agents/ppo/ppo.py

```diff
@@ -253,69 +170,137 @@ def warn_about_bad_reward_scales(config, result):
-                _fake_gpus=config["_fake_gpus"],
-                framework=config.get("framework")))
-
-    # Update KL after each round of training.
-    train_op = train_op.for_each(lambda t: t[1]).for_each(UpdateKL(workers))
-
-    # Warn about bad reward scales and return training metrics.
-    return StandardMetricsReporting(train_op, workers, config) \
-        .for_each(lambda result: warn_about_bad_reward_scales(config, result))
-
-
-# Build a child class of `Trainer`, which uses the framework specific Policy
-# determined in `get_policy_class()` above.
-PPOTrainer = build_trainer(
-    name="PPO",
-    default_config=DEFAULT_CONFIG,
-    validate_config=validate_config,
-    default_policy=PPOTFPolicy,
-    get_policy_class=get_policy_class,
-    execution_plan=execution_plan,
-)
+class PPOTrainer(Trainer):
+    @classmethod
+    @override(Trainer)
+    def get_default_config(cls) -> TrainerConfigDict:
+        return DEFAULT_CONFIG
+
+    @override(Trainer)
+    def validate_config(self, config: TrainerConfigDict) -> None:
+        """Validates the Trainer's config dict.
+
+        Args:
+            config (TrainerConfigDict): The Trainer's config to check.
+
+        Raises:
+            ValueError: In case something is wrong with the config.
+        """
+        if isinstance(config["entropy_coeff"], int):
+            config["entropy_coeff"] = float(config["entropy_coeff"])
+
+        if config["entropy_coeff"] < 0.0:
+            raise DeprecationWarning("entropy_coeff must be >= 0.0")
+
+        # SGD minibatch size must be smaller than train_batch_size (b/c
+        # we subsample a batch of `sgd_minibatch_size` from the train-batch for
+        # each `sgd_num_iter`).
+        # Note: Only check this if `train_batch_size` > 0 (DDPPO sets this
+        # to -1 to auto-calculate the actual batch size later).
+        if config["train_batch_size"] > 0 and \
+                config["sgd_minibatch_size"] > config["train_batch_size"]:
+            raise ValueError("`sgd_minibatch_size` ({}) must be <= "
+                             "`train_batch_size` ({}).".format(
+                                 config["sgd_minibatch_size"],
+                                 config["train_batch_size"]))
+
+        # Check for mismatches between `train_batch_size` and
+        # `rollout_fragment_length` and auto-adjust `rollout_fragment_length`
+        # if necessary.
+        # Note: Only check this if `train_batch_size` > 0 (DDPPO sets this
+        # to -1 to auto-calculate the actual batch size later).
+        num_workers = config["num_workers"] or 1
+        calculated_min_rollout_size = \
+            num_workers * config["num_envs_per_worker"] * \
+            config["rollout_fragment_length"]
+        if config["train_batch_size"] > 0 and \
+                config["train_batch_size"] % calculated_min_rollout_size != 0:
+            new_rollout_fragment_length = config["train_batch_size"] // (
```

Comment:
high level question, should we do this for all on-policy agents?
train batch size is such a mysterious thing for us.
logics like this living in a specific agent makes things less consistent.
I know some agents don't work like this, but for the ones do, should we put this in a util function, so they can all do this at the beginning of a run.

definitely not belong to this PR, just curious what you think.

Reply:
I completely agree with you that train batch sizes should be handled not by individual trainers, but in a more generic way, as we discussed offline. This just bubbled up here b/c I had to move that block of code. But yeah, we should make a separate PR in which we implement "guaranteed batch sizes" for all algos.